.github/workflows/build-binaries.yml

```diff
@@ -29,46 +33,67 @@ jobs:
         with:
           fetch-depth: 0
 
+      - name: Set up Go
+        uses: actions/setup-go@v5
+        with:
+          go-version-file: go.mod
+          cache: true
+
+      - name: Set up Rust
+        uses: dtolnay/rust-toolchain@stable
+      
+      - name: Cache Rust dependencies
+        uses: Swatinem/rust-cache@v2
+        with:
+          workspaces: |
+            vm/rust
+            core/rust
+            starknet/compiler/rust
+
       - name: Get latest tag
         run: echo "TAG=$(git describe --tags)" >> $GITHUB_ENV
 
       - name: Get artifact name
-        run: echo "ARTIFACT_NAME=juno-${{ env.TAG }}-${{ runner.os }}-$(uname -m)" >> $GITHUB_ENV
+        run: |
+          OS_NAME=$([ "${{ runner.os }}" == "macOS" ] && echo "darwin" || echo "linux")
+          echo "ARTIFACT_NAME=juno-${{ env.TAG }}-${OS_NAME}-${{ matrix.arch }}" >> $GITHUB_ENV
 
       - name: Install dependencies (Linux)
         if: runner.os == 'Linux'
-        run: sudo apt-get update -qq && sudo apt-get install -y upx-ucl build-essential cargo git golang libjemalloc-dev libjemalloc2 -y
+        run: |
+          sudo apt-get update -qq
+          sudo apt-get install -y upx-ucl build-essential libjemalloc-dev libjemalloc2 libbz2-dev
 
       - name: Install dependencies (macOS)
         if: runner.os == 'macOS'
-        run: brew install cargo-c jemalloc
-
-      - name: Set up Go
-        uses: actions/setup-go@v5
-        with:
-          go-version-file: go.mod
+        run: brew install jemalloc
           
-      - name: Build Juno
+      - name: Build binary
+        run: make juno
+
+      - name: Compress binary (Linux)
+        if: runner.os == 'Linux'
         run: |
-          make juno
-          if [[ "${{ runner.os }}" != "macOS" ]]; then
-            upx build/juno
-          fi
+          upx build/juno
```

Comment:
can you remind me why we don't use `upx` for macOS ?

Reply:
UPX and macOS don’t always work nice. Apple’s Gatekeeper can flag UPX-packed binaries as suspicious, and macOS already compresses binaries, so using UPX doesn’t add much. Instead, it might just cause more headaches than it’s worth.

Reply:
thanks 🙇